File: benchmarks/kmeans/kmeans-scala/kmeans.scala
import org.apache.spark.mllib.clustering.{KMeans, KMeansModel}
import org.apache.spark.mllib.linalg.Vectors
import org.apache.spark.SparkContext
import org.apache.spark.SparkConf
import org.apache.spark.sql.Row


object benchKmeans {
  def main(args: Array[String]) {

    val conf       = new SparkConf().setAppName("K-means Benchmark")
    val sc         = new SparkContext(conf)
    val sqlContext = new org.apache.spark.sql.SQLContext(sc)

    var json_file = "reviews_books_first_1000.json"
    if(args.length == 2){
      json_file = args(1);
    }
    println("Reading from file: " + json_file);

    // Create the DataFrame
    val df = sqlContext.read.json(json_file)

    val parsedData = df.map(t => Vectors.dense(
                              t.getAs[Double]("overall"),
                              t.getAs[String]("reviewText").length().toDouble,
                              t.getAs[Long]("unixReviewTime").toDouble
                                              )).cache()

    // Cluster the data into two classes using KMeans
    val numClusters   = 2
    val numIterations = 10
    val numRuns = 10
    val initMode = "random"
    val clusters      = KMeans.train(parsedData, numClusters, numIterations, numRuns, initMode)

    // Evaluate clustering by computing Within Set Sum of Squared Errors
    val WSSSE = clusters.computeCost(parsedData)
      println("Within Set Sum of Squared Errors = " + WSSSE)
      clusters.clusterCenters.map(x => println(x))

    // Save and load model
    clusters.save(sc, "myModelPath")
    val sameModel = KMeansModel.load(sc, "myModelPath")
  }
}

File: benchmarks/kmeans/kmeans-python/kmeans.py
from pyspark.mllib.clustering import KMeans, KMeansModel
# from numpy import array
from math import sqrt
from pyspark import SparkContext
import json

# Load and parse the data
sc = SparkContext("local", "Python K-Means Amazon Reviews")

data = sc.textFile('reviews_books_first_1000.json')
parsedData = data.map(lambda line: json.loads(line)).\
             map(lambda line: (line['overall'], len(line['reviewText'])))

# Build the model (cluster the data)
clusters = KMeans.train(parsedData, 2, maxIterations=10,
                        runs=10, initializationMode="random")


# Evaluate clustering by computing Within Set Sum of Squared Errors
def error(point):
    center = clusters.centers[clusters.predict(point)]
    return sqrt(sum([x**2 for x in (point - center)]))

WSSSE = parsedData.map(lambda point: error(point)).reduce(lambda x, y: x + y)
print("Within Set Sum of Squared Error = " + str(WSSSE))

# Save and load model
clusters.save(sc, "myModelPath")
sameModel = KMeansModel.load(sc, "myModelPath")
